java/src/main/native/src/row_conversion.cu

```diff
@@ -923,106 +910,89 @@ __global__ void copy_from_rows(const size_type num_rows, const size_type num_col
   // to speed up some of the random access memory we do, we copy col_sizes and col_offsets
   // to shared memory for each of the tiles that we work on
 
-  constexpr unsigned stages_count = NUM_TILES_PER_KERNEL_LOADED;
-  auto group = cooperative_groups::this_thread_block();
-  extern __shared__ int8_t shared_data[];
-  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_tile};
+  auto const group = cooperative_groups::this_thread_block();
+  auto const warp = cooperative_groups::tiled_partition<cudf::detail::warp_size>(group);
+  extern __shared__ int8_t shared[];
 
 #ifdef ASYNC_MEMCPY_SUPPORTED
   // Initialize cuda barriers for each tile.
-  __shared__ cuda::barrier<cuda::thread_scope_block> tile_barrier[NUM_TILES_PER_KERNEL_LOADED];
+  __shared__ cuda::barrier<cuda::thread_scope_block> tile_barrier;
   if (group.thread_rank() == 0) {
-    for (int i = 0; i < NUM_TILES_PER_KERNEL_LOADED; ++i) {
-      init(&tile_barrier[i], group.size());
-    }
+    init(&tile_barrier, group.size());
   }
   group.sync();
 #endif // ASYNC_MEMCPY_SUPPORTED
 
-  auto tiles_remaining =
-      std::min(static_cast<uint>(tile_infos.size()) - blockIdx.x * NUM_TILES_PER_KERNEL_FROM_ROWS,
-               static_cast<uint>(NUM_TILES_PER_KERNEL_FROM_ROWS));
-
-  size_t fetch_index;
-  size_t processing_index;
-  for (processing_index = fetch_index = 0; processing_index < tiles_remaining; ++processing_index) {
-    // Fetch ahead up to stages_count groups
-    for (; fetch_index < static_cast<size_t>(tiles_remaining) &&
-           fetch_index < (processing_index + stages_count);
-         ++fetch_index) {
-      auto const fetch_tile = tile_infos[blockIdx.x * NUM_TILES_PER_KERNEL_FROM_ROWS + fetch_index];
-      auto const fetch_tile_start_row = fetch_tile.start_row;
-      auto const starting_col_offset = col_offsets[fetch_tile.start_col];
-      auto const fetch_tile_row_size = fetch_tile.get_shared_row_size(col_offsets, col_sizes);
-      auto const row_batch_start =
-          fetch_tile.batch_number == 0 ? 0 : batch_row_boundaries[fetch_tile.batch_number];
+  {
+    auto const fetch_tile = tile_infos[blockIdx.x];
+    auto const fetch_tile_start_row = fetch_tile.start_row;
+    auto const starting_col_offset = col_offsets[fetch_tile.start_col];
+    auto const fetch_tile_row_size = fetch_tile.get_shared_row_size(col_offsets, col_sizes);
+    auto const row_batch_start =
+        fetch_tile.batch_number == 0 ? 0 : batch_row_boundaries[fetch_tile.batch_number];
+
+    for (int absolute_row = warp.meta_group_rank() + fetch_tile.start_row;
+         absolute_row <= fetch_tile.end_row; absolute_row += warp.meta_group_size()) {
+      warp.sync();
+      auto shared_offset = (absolute_row - fetch_tile_start_row) * fetch_tile_row_size;
+      auto dst = &shared[shared_offset];
+      auto src = &input_data[row_offsets(absolute_row, row_batch_start) + starting_col_offset];
```

Comment:
Hah! I just realized that `RowOffsetFunctor` might be a more appropriate name for `RowOffsetIter`.

Reply:
Although I read it as `Iter", I kept converting to "Functor" subconsciously. 
Thank you for accommodating by changing the name.